sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala

```diff
@@ -599,10 +599,39 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] with AliasHelpe
         if (Utils.isTesting) {
           assert(mayHaveCountBug.isDefined)
         }
+
+        def queryOutputFoldable(list: Seq[NamedExpression]): Boolean = {
+          trimAliases(list.filter(p => p.exprId.equals(query.output.head.exprId)).head).foldable
+        }
+
+        // SPARK-43156: We can judge whether the column returned by subquery is
+        // foldable (already handle by [[NullPropagation]]). If it is, it means that
+        // the result of this value has no substantial relationship with the data,
+        // and the presence or absence of data will not affect this column. So in
+        // this case, this column can be extracted from the JOIN to ensure that this
+        // value can be obtained regardless of whether the data JOIN is successful or not.
+        lazy val resultFoldable = {
+          query match {
+            case Project(expressions, _) =>
+              queryOutputFoldable(expressions)
+            case Aggregate(_, expressions, _) =>
+              queryOutputFoldable(expressions)
+            case _ =>
+              false
+          }
+        }
+
         if (resultWithZeroTups.isEmpty) {
           // CASE 1: Subquery guaranteed not to have the COUNT bug because it evaluates to NULL
           // with zero tuples.
           planWithoutCountBug
+        } else if (mayHaveCountBug.getOrElse(false) && resultFoldable &&
```

Comment:
can you show me which test will fail if we don't have the changes here?

Reply:
Test case: 
<img width="625" alt="image" src="https://user-images.githubusercontent.com/32387433/234505362-b17db89d-7f27-4c62-979f-b39238f1ed6b.png">
Remove `resultFoldable`: 
<img width="399" alt="image" src="https://user-images.githubusercontent.com/32387433/234505527-fc0f6d2e-85d6-4dfb-82d4-2821aa92339c.png">
Should be:
<img width="398" alt="image" src="https://user-images.githubusercontent.com/32387433/234505644-a9b6de27-cdf3-4ec1-90ce-1c087efcb2f6.png">

Reply:
> can you show me which test will fail if we don't have the changes here?

if without this, the logic will use `Literal(0)` as result to replace `count(1)`.

Reply:
are you using the latest Spark version? It works fine if I only fix `l.copy(value = null)`
```
scala> sql("SELECT (SELECT count(1) FROM t2 WHERE t1.c1 = t2.c2) FROM t1").show
+------------------+
|scalarsubquery(c1)|
+------------------+
|                 1|
|                 0|
+------------------+
```

Reply:
Yes, after I merge the code. The test will passed. But the plan will be
```
== Optimized Logical Plan ==
Project [if (isnull(alwaysTrue#253)) 0 else count(1)#250L AS scalarsubquery(c2)#251L]
+- Join LeftOuter, (c1#247 = c2#245)
   :- Project [col1#246 AS c2#245]
   :  +- LocalRelation [col1#246]
   +- Aggregate [c1#247], [count(1) AS count(1)#250L, c1#247, true AS alwaysTrue#253]
      +- Project [col1#248 AS c1#247]
         +- LocalRelation [col1#248]
```
not
```
== Optimized Logical Plan ==
Project [false AS scalarsubquery(c2)#252]
+- LocalRelation [col1#247]
```

Reply:
Should I split it in two PR? Or just chagne title and description?

Reply:
There are some change after I create this PR. Seem like `resultFoldable` are useful when I create PR but not now.